Content.Server/Ninja/Systems/NinjaGlovesSystem.cs

```diff
@@ -0,0 +1,304 @@
+    [Dependency] private readonly DoAfterSystem _doafter = default!;
+    [Dependency] private readonly ElectrocutionSystem _electrocution = default!;
+    [Dependency] private readonly EmagSystem _emag = default!;
+    [Dependency] private readonly GameTicker _gameTicker = default!;
+    [Dependency] private readonly NinjaSystem _ninja = default!;
+    [Dependency] private readonly IPrototypeManager _proto = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly TagSystem _tags = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<ActivateInWorldEvent>(OnActivate);
+        SubscribeLocalEvent<NinjaGlovesComponent, DoAfterEvent>(EndedDoAfter);
+        // TODO: raising 1 event triggers for all 3 components sicne on same entity, change to be per-component?
+        SubscribeLocalEvent<NinjaDrainComponent, DoAfterEvent>(OnDrainDoAfter);
+        SubscribeLocalEvent<NinjaDownloadComponent, DoAfterEvent>(OnDownloadDoAfter);
+        SubscribeLocalEvent<NinjaTerrorComponent, DoAfterEvent>(OnTerrorDoAfter);
+
+        SubscribeLocalEvent<DoorComponent, DoorEmaggedEvent>(OnDoorEmagged);
+    }
+
+    protected override void NinjaEquippedGloves(EntityUid uid, NinjaGlovesComponent comp, EntityUid user, NinjaComponent ninja)
+    {
+        base.NinjaEquippedGloves(uid, comp, user, ninja);
+
+        if (TryComp<ActionsComponent>(user, out var actions))
+            _actions.AddAction(user, comp.ToggleAction, uid, actions);
+    }
+
+    protected override void UserUnequippedGloves(EntityUid uid, NinjaGlovesComponent comp, EntityUid user)
+    {
+        base.UserUnequippedGloves(uid, comp, user);
+
+        _actions.RemoveProvidedActions(user, uid);
+    }
+
+    // TODO: make this per-ability
+    private void OnActivate(ActivateInWorldEvent args)
+    {
+        var user = args.User;
+        var target = args.Target;
+        if (args.Handled
+            || !TryComp<NinjaComponent>(user, out var ninja)
+            || ninja.Gloves == null
+            || !TryComp<NinjaGlovesComponent>(ninja.Gloves, out var comp)
+            || !comp.Enabled)
+            return;
+
+        var uid = ninja.Gloves.Value;
+
+        // doorjack ability
+        if (TryComp<NinjaDoorjackComponent>(uid, out var doorjack) && HasComp<DoorComponent>(target))
+        {
+            if (_tags.HasTag(target, doorjack.EmagImmuneTag))
+                return;
+
+            var handled = _emag.DoEmagEffect(user, target);
+            if (!handled)
+                return;
+
+            _popups.PopupEntity(Loc.GetString("emag-success", ("target", Identity.Entity(target, EntityManager))), user,
+                user, PopupType.Medium);
+            _adminLogger.Add(LogType.Emag, LogImpact.High, $"{ToPrettyString(user):player} doorjacked {ToPrettyString(target):target}");
+
+            args.Handled = true;
+            return;
+        }
```

Comment:
```suggestion
    private void OnInteractAttempt(EntityUid uid, NinjaGlovesComponent component, InteractionAttemptEvent args)
    {
        if (!component.Enabled || args.Target == null)
            return;

        // doorjack ability
        if (TryComp<NinjaDoorjackComponent>(uid, out var doorjack) && HasComp<DoorComponent>(args.Target))
        {
            if (_tags.HasTag(args.Target.Value, doorjack.EmagImmuneTag))
                return;

            var handled = _emag.DoEmagEffect(uid, args.Target.Value);
            if (!handled)
                return;

            _popups.PopupEntity(Loc.GetString("emag-success", ("target", Identity.Entity(args.Target.Value, EntityManager))), uid,
                uid, PopupType.Medium);
            _adminLogger.Add(LogType.Emag, LogImpact.High, $"{ToPrettyString(uid):player} doorjacked {ToPrettyString(args.Target.Value):target}");
        }
    }
```

Take this out of the ActivateInWorld event and put it into the OnInteractAttempt method (from the InteractAttemptEvent from earlier).

Though honesty these extra components should be removed but I'll get to that review a bit later.